// @raven/cell/lib/index.js
import { __assign } from "tslib";
import ElementUI from 'element-ui';
import importComNames from './components/importComNames';
import exchange from './exchange';
var pkg = require('../package.json');
var _a = ElementUI, locale = _a.locale, i18n = _a.i18n;
var components = [];
function editComponentName(component) {
    if (component && component.name) {
        component.name = component.name.replace(/El/g, 'rv');
    }
}
Object.keys(ElementUI).forEach(function (comName) {
    var component = ElementUI[comName];
    if (importComNames.includes(comName) && component) {
        editComponentName(component);
        components.push(component);
    }
});
exchange(components);
var install = function (Vue, opts) {
    if (opts === void 0) { opts = {}; }
    locale(opts.locale);
    i18n(opts.i18n);
    components.forEach(function (component) {
        Vue.component(component.name, component);
    });
    Vue.use(ElementUI.Loading.directive);
    Vue.use(ElementUI.InfiniteScroll);
    Vue.prototype.$ELEMENT = {
        size: opts.size || '',
        zIndex: opts.zIndex || 2000
    };
    Vue.prototype.$loading = ElementUI.Loading.service;
    Vue.prototype.$msgbox = ElementUI.MessageBox;
    Vue.prototype.$alert = ElementUI.MessageBox.alert;
    Vue.prototype.$confirm = ElementUI.MessageBox.confirm;
    Vue.prototype.$prompt = ElementUI.MessageBox.prompt;
    Vue.prototype.$notify = ElementUI.Notification;
    Vue.prototype.$message = ElementUI.Message;
};
export * from 'element-ui';
export * from './locale/lang';
export default __assign(__assign({}, ElementUI), { version: pkg.version, locale: locale,
    i18n: i18n,
    install: install });
import Alert from './Alert';
import Aside from './Aside';
import Autocomplete from './Autocomplete';
import Avatar from './Avatar';
import Backtop from './Backtop';
import Badge from './Badge';
import Breadcrumb from './Breadcrumb';
import BreadcrumbItem from './BreadcrumbItem';
import Button from './Button';
import ButtonGroup from './ButtonGroup';
import Calendar from './Calendar';
import Card from './Card';
import Carousel from './Carousel';
import CarouselItem from './CarouselItem';
import Cascader from './Cascader';
// import CascaderPanel from './CascaderPanel'
import Checkbox from './Checkbox';
import CheckboxButton from './CheckboxButton';
import CheckboxGroup from './CheckboxGroup';
import Col from './Col';
import Collapse from './Collapse';
import CollapseItem from './CollapseItem';
import ColorPicker from './ColorPicker';
import Container from './Container';
import DatePicker from './DatePicker';
import Dialog from './Dialog';
import Divider from './Divider';
import Drawer from './Drawer';
import Dropdown from './Dropdown';
import DropdownItem from './DropdownItem';
import DropdownMenu from './DropdownMenu';
import Footer from './Footer';
import Form from './Form';
import FormItem from './FormItem';
import Header from './Header';
import Image from './Image';
import Input from './Input';
import InputNumber from './InputNumber';
import InfiniteScroll from './InfiniteScroll';
// import TimelineItem from './timeline-item';
import Link from './Link';
import Main from './Main';
import Menu from './Menu';
import MenuItem from './MenuItem';
import MenuItemGroup from './MenuItemGroup';
import Option from './Option';
import OptionGroup from './OptionGroup';
// import InfiniteScroll from './infinite-scroll';
import PageHeader from './PageHeader';
import Pagination from './Pagination';
import Popover from './Popover';
import Progress from './Progress';
import Radio from './Radio';
import RadioButton from './RadioButton';
import RadioGroup from './RadioGroup';
import Rate from './Rate';
import Row from './Row';
import Select from './Select';
import Slider from './Slider';
import Step from './Step';
import Steps from './Steps';
import Submenu from './Submenu';
import Switch from './Switch';
import Skeleton from './Skeleton';
import SkeletonItem from './SkeletonItem';
import Table from './Table';
import TableColumn from './TableColumn';
import TabPane from './TabPane';
import Tabs from './Tabs';
import Tag from './Tag';
import Timeline from './Timeline';
import TimePicker from './TimePicker';
import TimeSelect from './TimeSelect';
import Tooltip from './Tooltip';
import Transfer from './Transfer';
import Tree from './Tree';
import Upload from './Upload';
import Result from './Result';
import Empty from './Empty';
import Descriptions from './Descriptions';
// 按需导出
export { Pagination, Dialog, Autocomplete, Dropdown, DropdownMenu, DropdownItem, Menu, Submenu, MenuItem, MenuItemGroup, Input, InputNumber, Radio, RadioGroup, RadioButton, Checkbox, CheckboxButton, CheckboxGroup, Switch, Select, Option, OptionGroup, Button, ButtonGroup, Table, TableColumn, DatePicker, TimeSelect, TimePicker, Popover, Tooltip, Breadcrumb, BreadcrumbItem, Form, FormItem, Tabs, TabPane, Tag, Tree, Alert, Slider, Row, Col, Upload, Progress, Badge, Card, Rate, Steps, Step, Carousel, CarouselItem, Collapse, CollapseItem, Cascader, ColorPicker, Transfer, Container, Header, Aside, Main, Footer, Timeline, 
// TimelineItem,
Link, Divider, Image, Calendar, Backtop, InfiniteScroll, PageHeader, 
// CascaderPanel,
Avatar, Drawer, SkeletonItem, Skeleton, Result, Empty, Descriptions };
