// vite-plugin-utils/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KNOWN_SFC_EXTENSIONS = exports.JS_EXTENSIONS = exports.cleanUrl = exports.hashRE = exports.queryRE = exports.singlelineCommentsRE = exports.multilineCommentsRE = exports.OfficialPlugins = exports.sortPlugin = void 0;
var sort_plugin_1 = require("./sort-plugin");
Object.defineProperty(exports, "sortPlugin", { enumerable: true, get: function () { return sort_plugin_1.sortPlugin; } });
Object.defineProperty(exports, "OfficialPlugins", { enumerable: true, get: function () { return sort_plugin_1.OfficialPlugins; } });
exports.multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm;
exports.singlelineCommentsRE = /\/\/.*/g;
exports.queryRE = /\?.*$/s;
exports.hashRE = /#.*$/s;
const cleanUrl = (url) => url.replace(exports.hashRE, '').replace(exports.queryRE, '');
exports.cleanUrl = cleanUrl;
exports.JS_EXTENSIONS = [
    '.mjs',
    '.js',
    '.ts',
    '.jsx',
    '.tsx',
    '.cjs'
];
exports.KNOWN_SFC_EXTENSIONS = [
    '.vue',
    '.svelte',
];
